File: ruby/lib/voxgig_station/events.rb
# frozen_string_literal: true

# The solo event surface (design station.md 6): a bounded ring buffer plus
# a live tap with serialized callbacks. Events never fail an operation;
# overflow drops oldest and the drop count is visible in status.
#
# A port of typescript/src/events.ts, which is canonical. Public station
# operations are safe from any thread (design station.md 10.2), so the
# ring and tap list are guarded by a mutex; tap callbacks run serialized
# under it.

module VoxgigStation
  class EventBuffer
    def initialize(max = nil)
      @max = max.nil? ? 1000 : max
      @ring = []
      @drops = 0
      @taps = []
      @mutex = Mutex.new
    end

    def emit(ev)
      taps = nil
      @mutex.synchronize do
        @ring << ev
        if @ring.length > @max
          @ring.shift
          @drops += 1
        end
        taps = @taps.dup
      end
      # Serialized, and a raising tap must not fail the operation that
      # emitted the event.
      taps.each do |fn|
        begin
          fn.call(ev)
        rescue StandardError
          # deliberately swallowed
        end
      end
    end

    def events
      @mutex.synchronize { @ring.dup }
    end

    # Subscribe; returns an unsubscribe lambda.
    def tap(fn = nil, &blk)
      fn ||= blk
      @mutex.synchronize { @taps << fn }
      -> { @mutex.synchronize { @taps.delete(fn) } }
    end

    def status
      @mutex.synchronize { { 'buffered' => @ring.length, 'dropped' => @drops } }
    end
  end
end

File: python/voxgig_station/events.py
# The solo event surface (design station.md 6): a bounded ring buffer plus
# a live tap with serialized callbacks. Events never fail an operation;
# overflow drops oldest and the drop count is visible in status().
#
# A port of typescript/src/events.ts, which is canonical. Synchronized
# with one lock: all public station operations are safe to call from any
# thread (design station.md 10.2), and CPython list ops alone do not
# guarantee the ring/drop-count invariant.

import threading


class EventBuffer:
    def __init__(self, max_events=None):
        self._ring = []
        self._max = 1000 if max_events is None else max_events
        self._drops = 0
        self._taps = []
        self._lock = threading.Lock()

    def emit(self, ev):
        with self._lock:
            self._ring.append(ev)
            if len(self._ring) > self._max:
                self._ring.pop(0)
                self._drops += 1
            taps = list(self._taps)

        # Serialized, and a throwing tap must not fail the operation that
        # emitted the event.
        for fn in taps:
            try:
                fn(ev)
            except Exception:
                pass

    def events(self):
        with self._lock:
            return list(self._ring)

    def tap(self, fn):
        with self._lock:
            self._taps.append(fn)

        def untap():
            with self._lock:
                if fn in self._taps:
                    self._taps.remove(fn)

        return untap

    def status(self):
        with self._lock:
            return {'buffered': len(self._ring), 'dropped': self._drops}
